File: DiscoJava/DiscoVirtual/src/main/java/org/example/Pasta.java
package org.example;

import java.util.ArrayList;
import java.util.List;

public class Pasta implements ItemArmazenamento {

    private String nome;

    private Pasta pastaPai;

    private List<ItemArmazenamento> itens;

    public Pasta(String nome, Pasta pastaPai) {
        this.setNome(nome);
        this.setPastaPai(pastaPai);
        this.itens = new ArrayList<>();
    }

    public Pasta(String nome) {
        this.setNome(nome);
        this.itens = new ArrayList<>();
    }


    public String getNome() {
        return nome;
    }

    public void setNome(String nome) {
        this.nome = nome;
    }

    public Pasta getPastaPai() {
        return this.pastaPai;
    }

    public void setPastaPai(Pasta pastaPai) {
        this.pastaPai = pastaPai;
        pastaPai.add(this);
    }

    @Override
    public void apagar() {

        if(pastaPai == null){
            throw new IllegalArgumentException("Não é possível apagar a raiz");
        }

        List<ItemArmazenamento> copia = new ArrayList<>(this.itens);

        for (ItemArmazenamento item : copia) {
            if (item instanceof Arquivo) {
                this.remove(item);
            } else {
                item.apagar();
            }
        }

        if (itens.isEmpty()) {
            this.pastaPai.remove(this);
        }
    }

    @Override
    public Boolean add(ItemArmazenamento item) {
        return this.itens.add(item);
    }

    @Override
    public Boolean remove(ItemArmazenamento item) {
        return this.itens.remove(item);
    }

    @Override
    public List<ItemArmazenamento> getItens() {
        return this.itens;
    }

    @Override
    public long getTamanhoKB() {
        long tamanho = 0;
        for (ItemArmazenamento item : this.itens) {
            if(item instanceof Arquivo){
                tamanho += item.getTamanhoKB();
            }else{
                tamanho += item.getTamanhoKB();
            }
        }
        return tamanho;
    }

    @Override
    public String toString() {
        return "Pasta: " + this.getNome() + " " + this.getTamanhoKB() + "KB";
    }
}

File: DiscoTs/src/entities/Pasta.ts
import { Arquivo } from './Arquivo';
import { ItemArmazenamento } from './ItemArmazenamento';

class Pasta implements ItemArmazenamento {
  private nome: string;

  private itens: ItemArmazenamento[];

  private pastaPai: Pasta;

  constructor(nome: string, pastaPai?: Pasta) {
    this.setNome(nome);

    if (pastaPai) {
      this.setPastaPai(pastaPai);
    }

    this.itens = [];
  }

  public getNome(): string {
    return this.nome;
  }

  public setNome(nome: string): void {
    this.nome = nome;
  }

  public getPastaPai(): Pasta {
    return this.pastaPai;
  }

  private setPastaPai(pastaPai: Pasta): void {
    this.pastaPai = pastaPai;
  }

  async apagar(): Promise<void> {
    if (!this.pastaPai) {
      throw new Error('Não é possível apagar a pasta raiz');
    }

    const copia = [...this.itens];

    for (const item of copia) {
      if (item instanceof Arquivo) {
        await this.remove(item);
      } else {
        await item.apagar();
      }

      if (this.itens.length === 0) {
        await this.pastaPai.remove(this);
      }
    }
  }

  async getTamanhoKB(): Promise<number> {
    return 0;
  }

  async add(item: ItemArmazenamento): Promise<void> {
    this.itens.push(item);
  }

  async remove(item: ItemArmazenamento): Promise<void> {
    this.itens = this.itens.filter(i => i !== item);
  }

  async getItens(): Promise<ItemArmazenamento[]> {
    return this.itens;
  }
}

export { Pasta };
